lib/clerk/rack_middleware.rb

```diff
@@ -2,96 +2,186 @@
-        @error = e
+      def session
+        return nil if @token.nil?
+        return @session if @session
+
+        begin
+          @session = fetch_session
+        rescue Errors::Authentication => e
+          @error = e
+        end
+        @session
       end
-      @session
-    end
 
-    def user_id
-      @user_id ||= session.dig("user_id")
-    end
+      def user_id
+        @user_id ||= session.dig("user_id")
+      end
 
-    def user
-      return nil if session.nil?
-      @user ||= fetch_user(user_id)
-    end
+      def user
+        return nil if session.nil?
+        @user ||= fetch_user(user_id)
+      end
 
-    def debug
-      (instance_variables - [:@sdk]).map do |ivar|
-        [ivar.to_s, instance_variable_get(ivar)]
-      end.to_h
-    end
+      def debug
+        (instance_variables - [:@sdk]).map do |ivar|
+          [ivar.to_s, instance_variable_get(ivar)]
+        end.to_h
+      end
 
-    private
-    def sdk
-      @sdk ||= SDK.new
-    end
+      private
 
-    def cache_key
-      @cache_key ||= @token.split(".")[1]
-    end
+      def sdk
+        @sdk ||= SDK.new
+      end
 
-    def fetch_session
-      if session_id
-        cached_fetch("clerk_session:#{session_id}:#{cache_key}") do
-          sdk.sessions.verify_token(session_id, @token)
-        end
-      else
-        cached_fetch("clerk_session:#{cache_key}") do
-          client = sdk.clients.verify_token(@token)
-          @session_id = client["last_active_session_id"]
-          client["sessions"].find do |sess|
-            sess["id"] == @session_id
+      def cache_key
+        @cache_key ||= @token.split(".")[1]
+      end
+
+      def fetch_session
+        if session_id
+          cached_fetch("clerk_session:#{session_id}:#{cache_key}") do
+            sdk.sessions.verify_token(session_id, @token)
           end
+        else
+          cached_fetch("clerk_session:#{cache_key}") do
+            client = sdk.clients.verify_token(@token)
+            @session_id = client["last_active_session_id"]
+            client["sessions"].find do |sess|
+              sess["id"] == @session_id
+            end
+          end
+        end
+      end
+
+      def fetch_user(user_id)
+        cached_fetch("clerk_user:#{user_id}") do
+          sdk.users.find(user_id)
+        end
+      end
+
+      def cached_fetch(key, &block)
+        if store = Clerk.configuration.middleware_cache_store
+          store.fetch(key, expires_in: CACHE_TTL, &block)
+        else
+          yield
         end
       end
     end
 
-    def fetch_user(user_id)
-      cached_fetch("clerk_user:#{user_id}") do
-        sdk.users.find(user_id)
+    class ProxyNetworkless
+      attr_reader :session_id, :errors
+
+      def initialize(env)
+        @req = Rack::Request.new(env)
+
+        @session_id = @req.params["_clerk_session_id"]
+        @session = nil
+        @user_id = nil
+        @user = nil
+        @errors = []
+      end
+
+      def session
```

Comment:
```suggestion
      def session_token
```

Let's keep this as session_token, so `session` can use fapi to fetch the true Session object?

Reply:
Do we need a function in the middleware that fetches the session object? If so, can you elaborate on why? (Keep in mind that there's already the `clerk_session` helper method that we expose to users)

Note that such a rename will be a breaking change for users of the SDK that used that function.

Reply:
I think the main one is: Some users may not be comfortable with the minute of leeway and will want to double-check expiration with the network.

The secondary one we don’t have right now is: eventually the Session object may contain information that doesn’t exist on the token.

If we keep ‘session’ but it hits the network, while ‘session_token’ does not, will there actually be backward compatibility issues?

Reply:
> I think the main one is: Some users may not be comfortable with the minute of leeway and will want to double-check expiration with the network.

Reply:
Fair points, thanks for elaborating.

The fact that `session` doesn't map to a `Session` object is something that can be confusing as well, so that's another argument for changing it.

Truth is, this middleware already contains a backward-incompatible change (a middleware method rename from `error` to `errors`), so I think we'll anyway have inform customers that migrating to v2 will _conditionally_ require changes in the code as well.

Currently we provide the `clerk_session` helper. In AuthV1 this fetches the session object indeed, by querying the API. In AuthV2, `clerk_session` merely returns the session token. This in itself is another subtle backwards-incompatible change. So I don't think there's anyway to keep backwards compatibility between applications that are migrating from v1 to v2, regardless of what the developer does with those cookies. Depending on the actual use case we might always break some *customer* flows (not customers' user flows) when someone migrates from v1 to v2.

--------------------

Now to this specific point:

My suggestion, would be to add a `#fetch_clerk_session` helper (similar to `clerk_user`) that one can call from their controllers or views. The prefix "fetch" makes it somewhat more obvious that this involves a network roundtrip. 

Then we can rename the current `ProxyNetworkless#session` to `session_token`, which I agree is a more precise name, albeit a less important rename since most users won't need to use these middleware methods. So it's kind of orthogonal to how we name the aforementioned `#fetch_session` helper.

So:

- the internal middleware method is renamed from `#session` to `#session_token`
- users will be able to call `#fetch_clerk_session` (or `#clerk_fetch_session`). This will return a `Session` object, after validating that the token from `#session_token` is valid
- we leave `#clerk_session` helper intact, so as to preserve backwards compatibility. For authv2, it will return the session token (no roundtrips)
- we include all the required changes for users migrating from v1->v2 in the upgrade notes

WDYT?

Reply:
Hey - I think I need a little more help here to understand where we're losing backward compatibility.

I thought we need to retain backward compatibility to provide a zero-downtime upgrade path.  My impression was that the upgrade process would be:

1. Upgrade backend to an SDK that supports both authv1 and authv2
2. Update the frontend to authv2 after

What methods are the developer using today that we're changing for authv2?  Do we have this challenge in other SDKs?

Reply:
The process is what you described pretty much, with the caveat that in step (1) the developer might have to change some of its code as well, depending on how they use the SDK.

For instance, the middleware's `error` method has been renamed to `errors` and the return value is not an array instead of a string. Another breaking change is that, the `clerk_session` until now returned a `Session` object, where now it will return a session token.

So the steps involved will be to:

1. Upgrade backend to an SDK that supports both authv1 and authv2
2. Upgrade application code (if necessary) with the aforementioned changes
3. Update the frontend to authv2 after

Does this make sense?

Reply:
Note that this is still a zero-downtime upgrade path.